waku/waku_archive/driver/postgres_driver/postgres_driver.nim

```diff
@@ -587,6 +534,206 @@ proc sleep*(s: PostgresDriver, seconds: int):
+
+  self.futLoopPartitionFactory = self.loopPartitionFactory(onFatalError)
+
+
+proc removeOldestPartition(self: PostgresDriver,
+                           forceRemoval: bool = false, ## To allow cleanup in tests
+                           ):
+                           Future[ArchiveDriverResult[void]] {.async.}  =
+  ## Indirectly called from the retention policy
+
+  let oldestPartition = self.partitionMngr.getOldestPartition().valueOr:
+    return err("could not remove oldest partition: " & $error)
+
+  if not forceRemoval:
+    let now = times.now().toTime().toUnix()
+    let currentPartitionRes = self.partitionMngr.getPartitionFromDateTime(now)
+    if currentPartitionRes.isOk():
+      ## The database contains a partition that would store current messages.
+
+      if currentPartitionRes.get() == oldestPartition:
+        debug "Skipping to remove the current partition"
+        return ok()
+
+  ## In the following lines is where the partition removal happens.
+  ## Detach and remove the partition concurrently to not block the parent table (messages)
+  let detachPartitionQuery =
+    "ALTER TABLE messages DETACH PARTITION " & oldestPartition.getName() & " CONCURRENTLY;"
+  debug "removeOldestPartition", query = detachPartitionQuery
+  (await self.performWriteQuery(detachPartitionQuery)).isOkOr:
+    return err(fmt"error in {detachPartitionQuery}: " & $error)
+
+  ## Clear all data from the partition and retrieve back space to the operating system
+  let truncateQuery = "TRUNCATE TABLE " & oldestPartition.getName()
+  debug "removeOldestPartition truncate", query = truncateQuery
+  (await self.performWriteQuery(truncateQuery)).isOkOr:
+    return err(fmt"error in {truncateQuery}: " & $error)
+
+  ## Drop the partition
+  let dropPartitionQuery = "DROP TABLE " & oldestPartition.getName()
+  debug "removeOldestPartition drop partition", query = dropPartitionQuery
+  (await self.performWriteQuery(dropPartitionQuery)).isOkOr:
+    return err(fmt"error in {dropPartitionQuery}: " & $error)
+
+  self.partitionMngr.removeOldestPartitionName()
+
+  return ok()
+
+proc containsAnyPartition*(self: PostgresDriver): bool =
+  return not self.partitionMngr.isEmpty()
+
+method decreaseDatabaseSize*(driver: PostgresDriver,
+                             targetSizeInBytes: int64,
+                             forceRemoval: bool = false):
+                             Future[ArchiveDriverResult[void]] {.async.} =
+  var dbSize = (await driver.getDatabaseSize()).valueOr:
+    return err("decreaseDatabaseSize failed to get database size: " & $error)
+
+  ## database size in bytes
+  var totalSizeOfDB: int64 = int64(dbSize)
+
+  debug "reduce database size", targetSize = $targetSizeInBytes, currentSize = $totalSizeOfDB
+
+  if totalSizeOfDB <= targetSizeInBytes:
+    return ok()
+
+  while totalSizeOfDB > targetSizeInBytes and driver.containsAnyPartition():
+    (await driver.removeOldestPartition(forceRemoval)).isOkOr:
```

Comment:
In general it is an ok algorithm I think it might be to consider that at least collect information about peak time-ranges. Probably it will heavily depend on different use cases (what type of applications will use the store service) what will be the message count/size distribution among time periods we defined.
Would we consider to log somehow the size of the partitions (with time-period it contained)?

Reply:
> Would we consider to log somehow the size of the partitions (with time-period it contained)?

Good point! Tackled in https://github.com/waku-org/nwaku/commit/dce24e57858d370965cad4766f533bf8cf46397b

e.g.:
```
DBG 2024-03-06 20:12:26.948+01:00 removed partition                          tid=1981932 file=postgres_driver.nim:721 partition_name=messages_1709752346_1709755946 partition_size="32 kB"

```